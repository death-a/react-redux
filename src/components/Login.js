import { connect } from "react-redux";
import { useState } from "react";
import { validUsername, validPassword } from '../utils/Regex';
import { handleLogin } from "../actions/loggedinUser";

const Login = (props) => {
    const [signUp, setSignUp] = useState(false);
    const [username, setUsername] = useState("");
    const [name, setName] = useState("");
    const [values, setValues] = useState({
        password: "",
        showPassword: false,
    });
    const [errmsg, setErrMsg] = useState("");
    const [userSelected, setUserSelected] = useState("");

    const onLogIn = (e) => {
        e.preventDefault();
        if (username === "" && values.password === "") {
            setErrMsg("");
        }
        let matchUser = false;
        if (userSelected !== "" && username === userSelected) {
            props.dispatch(handleLogin(userSelected));
            matchUser = true;
        } else {
            const users = JSON.parse(localStorage.getItem("users"));
            if (users !== null) {
                for (const user of users) {
                    if (user["username"] === username && user["password"] !== values.password) {
                        setErrMsg('Username Password mismatch');
                        matchUser = true;
                    } else if (user["username"] === username && user["password"] === values.password) {
                        matchUser = true;
                    }
                }
            }
        }
        if (!matchUser) {
            setErrMsg("User does not exist in database. Please Sign Up!");
        }
    }

    const onSignUp = (e) => {
        e.preventDefault();
        let errUserName = false, errPass = false, errName = false, userExists = false;

        if (!validUsername.test(username)) {
            errUserName = true;
        }
        if (!validPassword.test(values.password)) {
            errPass = true;
        }
        if (name.trim() === "") {
            errName = true;
        }
        const users = JSON.parse(localStorage.getItem("users"));
        if (users !== null) {
            for (const user of users) {
                if (user["username"] === username) {
                    userExists = true;
                }
            }
        }
        if (!errUserName && !errPass && !errName && !userExists) {
            //handleSignUp(username, name, values.password);
        } else if (errUserName) {
            setErrMsg("Please check if all the rules for username are satisfied");
        } else if (errPass) {
            setErrMsg("Some of the rules for setting the password are not satisfied");
        } else if (errName) {
            setErrMsg("Name cannot be blank");
        } else if (userExists) {
            setErrMsg("Username already exists please enter a different username.");
        }
    }

    const handleClickShowPassword = () => {
        setValues({ ...values, showPassword: !values.showPassword });
    };

    const handleMouseDownPassword = (event) => {
        event.preventDefault();
    };

    const onhandleChangeUsername = (event) => {
        setUsername(event.target.value);
        setErrMsg("");
    }

    const onhandleChangeName = (event) => {
        setName(event.target.value);
        setErrMsg("");
    }

    const handlePasswordChange = (prop) => (event) => {
        setValues({ ...values, [prop]: event.target.value });
        setErrMsg("");
    };

    const handleUserSelection = (e) => {
        const selectVal = e.target.value;
        if (selectVal === "") {
            setUsername("");
            setName("");
            setValues({ password: "", showPassword: false });
            setErrMsg("");
        } else {
            setUsername(props.users[selectVal].id);
            setName(props.users[selectVal].name);
            setValues({ password: props.users[selectVal].password, showPassword: false });
            setErrMsg("");
        }
        setUserSelected(e.target.value);
    };

    return (
        <div>
            <p>For faster login use dropdownlist</p>
            <select onChange={handleUserSelection}>
                <option value="">Select a User to login</option>
                {Object.keys(props.users).map((id) => (
                    <option key={id} value={id}>{props.users[id].name}</option>
                ))}
            </select>
            <form >
                <div >
                    <p >{errmsg}</p>
                    <div >
                        <input type="text"
                            name="username"
                            placeholder="Username"
                            value={username}
                            onChange={(event) => onhandleChangeUsername(event)} />
                    </div>
                    {
                        (signUp) ?
                            <div >
                                <input type="text"
                                    name="name"
                                    placeholder="Name"
                                    value={name}
                                    onChange={(event) => onhandleChangeName(event)} />
                            </div>
                            :
                            ''
                    }
                    <div >
                        <input type={values.showPassword ? "text" : "password"}
                            name="password"
                            placeholder="Password"
                            value={values.password}
                            onChange={handlePasswordChange("password")} />

                        <span onClick={handleClickShowPassword}
                            onMouseDown={handleMouseDownPassword}
                        >
                            {
                                (values.showPassword) ? "Hide" : "Show"
                            }
                        </span>
                    </div>
                    {
                        (signUp) ?
                            <div >
                                <button name="signup"
                                    onClick={onSignUp}>
                                    Sign Up
                                </button>
                                <button name="olduser"
                                    onClick={() => { setSignUp(false); setErrMsg(""); }}>
                                    Already a User?
                                </button>
                            </div>
                            :
                            ""
                    }
                    {
                        (!signUp) ?
                            <div >
                                <button name="login"
                                    onClick={onLogIn}>
                                    Log In
                                </button>
                                <button name="newuser"
                                    onClick={() => { setSignUp(true); setErrMsg(""); }}>
                                    New User? Click here
                                </button>
                            </div>
                            :
                            ""
                    }
                    {
                        (signUp) ?
                            <div style={{ fontSize: "11px", color: "#737373", marginTop: "20px" }}>
                                <p><strong>Rules for Username:</strong><br></br>
                                    1. min 6 characters long<br></br>
                                    2. max 20 characters<br></br>
                                    3. allowed letters, numbers, dot and underscore.</p>

                                <p><strong>Rules for Password:</strong><br></br>
                                    1. min 4 characters<br></br>
                                    2. one smallcase letter and at least one number</p>
                            </div>
                            :
                            ""
                    }
                </div>
            </form>
        </div>
    )
}

const mapStateToProps = ({ users }) => ({
    users,
});

export default connect(mapStateToProps)(Login);