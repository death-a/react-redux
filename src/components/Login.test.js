import { createStore } from "redux";
import { Provider } from "react-redux";
import reducer from "../reducers";
import middleware from "../middleware";
import { render, fireEvent } from "@testing-library/react";
import Login from "./Login";

describe('Login', () => {
    /* Snapshot Testing */
    it('matches the snapshot when a user logs in', () => {
        const store = createStore(reducer, middleware);
        const component = render(
            <Provider store={store}>
                <Login />
            </Provider>);
        expect(component).toMatchSnapshot();
    });

    /* Login Test */
    it('will disable login button if username or password is blank', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: '' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: '' } });
        var logInButton = component.getByTestId('login-button');
        expect(logInButton).toBeDisabled();
    });

    it('will enable login button if username and password is filled', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: 'billiechan' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: 'pass123' } });
        var logInButton = component.getByTestId('login-button');
        expect(logInButton).not.toBeDisabled();
    });

    it('will display error message if username does not match any user', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: 'bil' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: 'pass123' } });
        var logInButton = component.getByTestId('login-button');
        fireEvent.click(logInButton);
        var errorMsg = component.queryByTestId('error-header');
        expect(errorMsg).toBeInTheDocument();
    });

    it('will display error message if password does not match username', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: 'billiechan' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: 'pass' } });
        var logInButton = component.getByTestId('login-button');
        fireEvent.click(logInButton);
        var errorMsg = component.queryByTestId('error-header');
        expect(errorMsg).toBeInTheDocument();
    });

    it('will allow user to login as existing user', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var newUserButton = component.getByTestId('newuser-button');
        fireEvent.click(newUserButton);
        var oldUserButton = component.getByTestId('olduser-button');
        fireEvent.click(oldUserButton);
        expect(component.getByTestId('login-button')).toBeInTheDocument();
        expect(component.queryByTestId('name-input')).not.toBeInTheDocument();
        expect(component.queryByTestId('signup-button')).not.toBeInTheDocument();
    });

    /* Sign Up Test */
    it('will allow user to sign up as new user', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var newUserButton = component.getByTestId('newuser-button');
        fireEvent.click(newUserButton);
        expect(component.queryByTestId('name-input')).toBeInTheDocument();
        expect(component.getByTestId('signup-button')).toBeInTheDocument();
        expect(component.queryByTestId('login-button')).not.toBeInTheDocument();
    });

    it('will disable sign up button if username, name or password is blank', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var newUserButton = component.getByTestId('newuser-button');
        fireEvent.click(newUserButton);
        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: '' } });
        var name = component.queryByTestId('name-input');
        fireEvent.change(name, { target: { value: '' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: '' } });
        var signUpButton = component.getByTestId('signup-button');
        expect(signUpButton).toBeDisabled();
    });

    it('will enable sign up button if username, name and password is filled', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var newUserButton = component.getByTestId('newuser-button');
        fireEvent.click(newUserButton);
        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: 'billiechan' } });
        var name = component.queryByTestId('name-input');
        fireEvent.change(name, { target: { value: 'Billie Chan' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: 'pass123' } });
        var signUpButton = component.getByTestId('signup-button');
        expect(signUpButton).not.toBeDisabled();
    });

    it('will display error message if username does not match criteria', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var newUserButton = component.getByTestId('newuser-button');
        fireEvent.click(newUserButton);
        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: 'bil' } });
        var name = component.queryByTestId('name-input');
        fireEvent.change(name, { target: { value: 'Billie Chan' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: 'pass123' } });
        var signUpButton = component.getByTestId('signup-button');
        fireEvent.click(signUpButton);
        var errorMsg = component.queryByTestId('error-header');
        var info = component.queryByTestId('criteria-info');
        expect(errorMsg).toBeInTheDocument();
        expect(info).toBeInTheDocument();
    });

    it('will display error message if password does not match criteria', () => {
        const store = createStore(reducer, middleware);
        var component = render(
            <Provider store={store}>
                <Login />
            </Provider>);

        var newUserButton = component.getByTestId('newuser-button');
        fireEvent.click(newUserButton);
        var username = component.queryByTestId('username-input');
        fireEvent.change(username, { target: { value: 'billiechan' } });
        var name = component.queryByTestId('name-input');
        fireEvent.change(name, { target: { value: 'Billie Chan' } });
        var password = component.queryByTestId('password-input');
        fireEvent.change(password, { target: { value: 'bil' } });
        var signUpButton = component.getByTestId('signup-button');
        fireEvent.click(signUpButton);
        var errorMsg = component.queryByTestId('error-header');
        var info = component.queryByTestId('criteria-info');
        expect(errorMsg).toBeInTheDocument();
        expect(info).toBeInTheDocument();
    });
})