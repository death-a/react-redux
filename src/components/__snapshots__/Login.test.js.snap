// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches the snapshot when a user logs in 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <p>
          For faster login use dropdownlist
        </p>
        <select
          data-testid="username-select"
        >
          <option
            value=""
          >
            Select a User to login
          </option>
        </select>
        <form>
          <div>
            <div>
              <input
                data-testid="username-input"
                name="username"
                placeholder="Username"
                type="text"
                value=""
              />
            </div>
            
            <div>
              <input
                data-testid="password-input"
                name="password"
                placeholder="Password"
                type="password"
                value=""
              />
              <span>
                Show
              </span>
            </div>
            
            <div>
              <button
                data-testid="login-button"
                disabled=""
                name="login"
              >
                Log In
              </button>
              <button
                data-testid="newuser-button"
                name="newuser"
              >
                New User? Click here
              </button>
            </div>
            
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <p>
        For faster login use dropdownlist
      </p>
      <select
        data-testid="username-select"
      >
        <option
          value=""
        >
          Select a User to login
        </option>
      </select>
      <form>
        <div>
          <div>
            <input
              data-testid="username-input"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
          </div>
          
          <div>
            <input
              data-testid="password-input"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
            <span>
              Show
            </span>
          </div>
          
          <div>
            <button
              data-testid="login-button"
              disabled=""
              name="login"
            >
              Log In
            </button>
            <button
              data-testid="newuser-button"
              name="newuser"
            >
              New User? Click here
            </button>
          </div>
          
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
